File: example/src/main/java/eu/darken/mvpbakery/example/screens/counting/CountingPresenter.java
package eu.darken.mvpbakery.example.screens.counting;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import javax.inject.Inject;

import eu.darken.mvpbakery.base.Presenter;
import eu.darken.mvpbakery.base.StateListener;
import eu.darken.mvpbakery.injection.ComponentPresenter;
import timber.log.Timber;


@CountingComponent.Scope
public class CountingPresenter extends ComponentPresenter<CountingPresenter.View, CountingComponent> implements StateListener {

    private final Counter counter;

    @Inject
    CountingPresenter(Counter counter) {
        this.counter = counter;
    }

    void onCountClick() {
        if (getView() != null) getView().showText(String.valueOf(counter.countUp()));
    }

    @Override
    public void onRestoreState(@Nullable Bundle bundle) {
        if (bundle != null) counter.setCounter(bundle.getInt("counter"));
    }

    @Override
    public void onBindChange(@Nullable View view) {
        super.onBindChange(view);
        if (view != null) view.showText(String.valueOf(counter.getCurrent()));
        onView(v -> v.showPresenterInfo(this));
    }

    @Override
    public void onSaveState(@NonNull Bundle bundle) {
        bundle.putInt("counter", counter.getCurrent());
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        Timber.d("onDestroy()");
    }

    public interface View extends Presenter.View {
        void showText(String text);

        void showPresenterInfo(CountingPresenter presenter);
    }
}

File: example/src/main/java/eu/darken/mvpbakery/example/screens/counting/CountingPresenter.kt
package eu.darken.mvpbakery.example.screens.counting

import android.os.Bundle
import eu.darken.mvpbakery.base.Presenter
import eu.darken.mvpbakery.base.StateListener
import eu.darken.mvpbakery.injection.ComponentPresenter
import timber.log.Timber
import javax.inject.Inject


@CountingComponent.Scope
class CountingPresenter @Inject
internal constructor(private val counter: Counter) : ComponentPresenter<CountingPresenter.View, CountingComponent>(), StateListener {

    fun onCountClick() {
        view?.showText(counter.countUp().toString())
    }

    override fun onRestoreState(inState: Bundle?) {
        if (inState != null) counter.setCounter(inState.getInt("counter"))
    }

    override fun onBindChange(view: View?) {
        super.onBindChange(view)
        view?.showText(counter.current.toString())
        withView { v -> v.showPresenterInfo(this) }
    }

    override fun onSaveState(outState: Bundle) {
        outState.putInt("counter", counter.current)
    }

    override fun onDestroy() {
        super.onDestroy()
        Timber.d("onDestroy()")
    }

    interface View : Presenter.View {
        fun showText(text: String)

        fun showPresenterInfo(presenter: CountingPresenter)
    }
}
